File: app/src/main/java/zmuzik/slidingpuzzle2/mainscreen/MainActivity.java
package zmuzik.slidingpuzzle2.mainscreen;


import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.StringTokenizer;

import javax.inject.Inject;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import zmuzik.slidingpuzzle2.App;
import zmuzik.slidingpuzzle2.Conf;
import zmuzik.slidingpuzzle2.R;
import zmuzik.slidingpuzzle2.common.Toaster;
import zmuzik.slidingpuzzle2.flickr.Photo;


public class MainActivity extends AppCompatActivity implements MainScreenView {

    final String TAG = this.getClass().getSimpleName();

    @BindView(R.id.viewPager)
    ViewPager mViewPager;
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.tabLayout)
    TabLayout mTabLayout;

    @BindString(R.string.title_section1)
    String mTitleSection1;
    @BindString(R.string.title_section2)
    String mTitleSection2;
    @BindString(R.string.title_section3)
    String mTitleSection3;
    @BindString(R.string.select_grid_size)
    String mSelectGridSize;
    @BindString(R.string.grid_size_selected_to)
    String mGridSizeSelectedTo;

    MainActivityComponent mComponent;
    WeakReference<SavedPicturesGridView> mSavedPicturesView;
    WeakReference<CameraPicturesGridView> mCameraPicturesView;
    WeakReference<FlickrPicturesGridView> mFlickrPicturesView;

    @Inject
    MainScreenPresenter mPresenter;
    @Inject
    Toaster mToaster;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        inject();
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        setSupportActionBar(mToolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);

        mViewPager.setAdapter(new ViewPagerAdapter());
        mViewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(mTabLayout));
        mTabLayout.setTabGravity(TabLayout.GRAVITY_FILL);
        mTabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                mViewPager.setCurrentItem(tab.getPosition());
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        mPresenter.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mPresenter.onPause();
    }

    private void inject() {
        mComponent = DaggerMainActivityComponent.builder()
                .appComponent(((App) getApplication()).getComponent(this))
                .mainActivityModule(new MainActivityModule(this))
                .build();
        mComponent.inject(this);
        mComponent.inject(mPresenter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_change_grid_size:
                openChangeGridSizeDialog();
                return true;
            case R.id.action_toggle_display_numbers:
                toggleShowNumbers();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void toggleShowNumbers() {
        boolean onOff = mPresenter.toggleShowNumbers();
        mToaster.show(onOff ? R.string.display_tile_numbers_on : R.string.display_tile_numbers_off);
    }

    private void openChangeGridSizeDialog() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(mSelectGridSize);
        builder.setSingleChoiceItems(Conf.GRID_SIZES, getGridDimsPosition(), new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int item) {
                String positionsStr = Conf.GRID_SIZES[item];
                StringTokenizer tokenizer = new StringTokenizer(positionsStr, "x");
                String shorterStr = tokenizer.nextToken();
                String longerStr = tokenizer.nextToken();
                mPresenter.setGridDimensions(Integer.parseInt(shorterStr), Integer.parseInt(longerStr));
                dialog.dismiss();
                mToaster.show(String.format(mGridSizeSelectedTo, positionsStr));
            }
        });
        builder.show();
    }

    private int getGridDimsPosition() {
        String currentDims = mPresenter.getGridDimensions();
        for (int i = 0; i < Conf.GRID_SIZES.length; i++) {
            if (currentDims.equals(Conf.GRID_SIZES[i])) return i;
        }
        return 0;
    }

    public MainActivityComponent getComponent() {
        return mComponent;
    }

    @Override
    public void updateSavedPictures(List<String> pictures) {
        if (mSavedPicturesView != null && mSavedPicturesView.get() != null) {
            mSavedPicturesView.get().update(pictures);
        }
    }

    @Override
    public void updateCameraPictures(List<String> pictures) {
        if (mCameraPicturesView != null && mCameraPicturesView.get() != null) {
            mCameraPicturesView.get().update(pictures);
        }
    }

    @Override
    public void updateFlickrPictures(List<Photo> photos) {
        if (mFlickrPicturesView != null && mFlickrPicturesView.get() != null) {
            mFlickrPicturesView.get().updatePhotos(photos);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        mPresenter.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private class ViewPagerAdapter extends PagerAdapter {
        @Override
        public Object instantiateItem(ViewGroup collection, int position) {
            BasePicturesGridView gridView = null;
            switch (position) {
                case 0:
                    gridView = new SavedPicturesGridView(MainActivity.this);
                    mSavedPicturesView = new WeakReference<SavedPicturesGridView>
                            ((SavedPicturesGridView) gridView);
                    break;
                case 1:
                    gridView = new CameraPicturesGridView(MainActivity.this);
                    mCameraPicturesView = new WeakReference<CameraPicturesGridView>
                            ((CameraPicturesGridView) gridView);
                    break;
                case 2:
                    gridView = new FlickrPicturesGridView(MainActivity.this);
                    mFlickrPicturesView = new WeakReference<FlickrPicturesGridView>
                            ((FlickrPicturesGridView) gridView);
                    break;
            }
            gridView.requestUpdate();
            collection.addView(gridView);
            return gridView;
        }

        @Override
        public void destroyItem(ViewGroup collection, int position, Object view) {
            collection.removeView((View) view);
        }

        @Override
        public int getCount() {
            return 3;
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }
    }
}

File: app/src/main/java/zmuzik/slidingpuzzle2/mainscreen/MainActivity.kt
package zmuzik.slidingpuzzle2.mainscreen


import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.view.PagerAdapter
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import kotlinx.android.synthetic.main.activity_main.*
import zmuzik.slidingpuzzle2.App
import zmuzik.slidingpuzzle2.Conf
import zmuzik.slidingpuzzle2.R
import zmuzik.slidingpuzzle2.common.Toaster
import zmuzik.slidingpuzzle2.common.di.ActivityScope
import zmuzik.slidingpuzzle2.flickr.Photo
import java.util.*
import javax.inject.Inject

@ActivityScope
class MainActivity : AppCompatActivity(), MainScreenView {

    internal val TAG = this.javaClass.simpleName

    lateinit var component: MainActivityComponent
    lateinit var toggleNumbersMenuItem: MenuItem

    lateinit var savedPicTab: SavedPicturesGridView
    lateinit var cameraPicTab: CameraPicturesGridView
    lateinit var flickrPicTab: FlickrPicturesGridView

    @Inject
    lateinit var presenter: MainScreenPresenter
    @Inject
    lateinit var toaster: Toaster

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        inject()
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)

        viewPager.adapter = ViewPagerAdapter()
        viewPager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabLayout))
        tabLayout.tabGravity = TabLayout.GRAVITY_FILL
        tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                viewPager.currentItem = tab.position
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}

            override fun onTabReselected(tab: TabLayout.Tab) {}
        })
    }

    override fun onResume() {
        super.onResume()
        presenter.onResume()
    }

    override fun onPause() {
        super.onPause()
        presenter.onPause()
    }

    private fun inject() {
        component = DaggerMainActivityComponent.builder()
                .appComponent((application as App).getComponent(this))
                .mainActivityModule(MainActivityModule(this))
                .build()
        component.inject(this)
        component.inject(presenter)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        toggleNumbersMenuItem = menu.findItem(R.id.action_toggle_display_numbers)
        setTileNumbersIcon(presenter.isShowTileNumbers())
        return true
    }

    internal fun setTileNumbersIcon(showNumbers: Boolean) {
        toggleNumbersMenuItem.setIcon(
                if (showNumbers)
                    R.drawable.ic_tile_with_number
                else
                    R.drawable.ic_tile_without_number)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_change_grid_size -> {
                openChangeGridSizeDialog()
                return true
            }
            R.id.action_toggle_display_numbers -> {
                toggleShowNumbers()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun toggleShowNumbers() {
        val onOff = presenter.toggleShowNumbers()
        setTileNumbersIcon(onOff)
        toaster.show(if (onOff) R.string.display_tile_numbers_on else R.string.display_tile_numbers_off)
    }

    private fun openChangeGridSizeDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(resources.getString(R.string.select_grid_size))
        builder.setSingleChoiceItems(Conf.GRID_SIZES, gridDimsPosition) { dialog, item ->
            val positionsStr = Conf.GRID_SIZES[item]
            val tokenizer = StringTokenizer(positionsStr, "x")
            val shorterStr = tokenizer.nextToken()
            val longerStr = tokenizer.nextToken()
            presenter.setGridDimensions(Integer.parseInt(shorterStr), Integer.parseInt(longerStr))
            dialog.dismiss()
            toaster.show(String.format(resources.getString(R.string.grid_size_selected_to), positionsStr))
        }
        builder.show()
    }

    private val gridDimsPosition: Int
        get() {
            val currentDims = presenter.getGridDimensions()
            return Conf.GRID_SIZES.indices.firstOrNull { currentDims == Conf.GRID_SIZES[it] } ?: 0
        }

    override fun updateSavedPictures(pictures: List<String>) {
        savedPicTab.update(pictures)
    }

    override fun setWaitingForCameraPictures() {
        cameraPicTab.setWaitingForPictures()
    }

    override fun updateCameraPictures(pictures: List<String>) {
        cameraPicTab.update(pictures)
    }

    override fun setWaitingForFlickrPictures() {
        flickrPicTab.setWaitingForPictures()
    }

    override fun updateFlickrPictures(photos: List<Photo>) {
        flickrPicTab.updatePhotos(photos)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        presenter.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun isCameraTabActive() = viewPager.currentItem == 1

    private inner class ViewPagerAdapter : PagerAdapter() {

        override fun instantiateItem(views: ViewGroup, position: Int): Any {
            val gridView: BasePicturesGridView = when (position) {
                0 -> {
                    savedPicTab = SavedPicturesGridView(this@MainActivity)
                    savedPicTab
                }
                1 -> {
                    cameraPicTab = CameraPicturesGridView(this@MainActivity)
                    cameraPicTab
                }
                else -> {
                    flickrPicTab = FlickrPicturesGridView(this@MainActivity)
                    flickrPicTab
                }
            }
            gridView.requestUpdate()
            views.addView(gridView)
            return gridView
        }

        override fun destroyItem(views: ViewGroup, pos: Int, view: Any) = views.removeView(view as View)

        override fun getCount() = 3

        override fun isViewFromObject(view: View, obj: Any) = view === obj
    }
}
